ios/brave-ios/Sources/Brave/Frontend/Brave Translate/BraveTranslateTabHelper.swift

```diff
@@ -165,17 +258,84 @@ class BraveTranslateTabHelper: NSObject {
   }
 
   @MainActor
-  func setupTranslate() async throws {
-    guard let tab = tab else {
+  func setupOnboarding() async throws {
+    guard tab != nil else {
+      return
+    }
+
+    if delegate?.canShowTranslateOnboarding(tab: tab) == true {
+      await detectLanguage()
+    }
+  }
+
+  @MainActor
+  func beginSetup() async throws {
+    guard tab != nil else {
       return
     }
 
     isTranslationReady = true
-    let languageInfo = await getLanguageInfo(for: tab)
-    self.currentLanguageInfo.pageLanguage = languageInfo.pageLanguage
-    self.currentLanguageInfo.currentLanguage = languageInfo.currentLanguage
+    await detectLanguage()
+  }
+
+  @MainActor
+  func finishSetup() async throws {
+    guard let tab = tab, let pageLanguage = currentLanguageInfo.pageLanguage else {
+      delegate?.updateTranslateURLBar(tab: tab, state: .unavailable)
+      return
+    }
+
+    // Check if the translation language is supported
+    let isTranslationSupported = await BraveTranslateSession.isTranslationSupported(
+      from: pageLanguage,
+      to: currentLanguageInfo.currentLanguage
+    )
+
+    delegate?.updateTranslateURLBar(
+      tab: tab,
+      state: isTranslationSupported ? .available : .unavailable
+    )
+
+    // Translation is not supported on this page, so do not show onboarding
+    // Return immediately
+    if !isTranslationSupported {
+      return
+    }
+
     try Task.checkCancellation()
-    try await updateTranslationStatus(for: tab)
+
+    // Check if the user can view the translation onboarding
+    guard delegate?.canShowTranslateOnboarding(tab: tab) == true else {
+      let translateEnabled = Preferences.Translate.translateEnabled.value == true
+
+      delegate?.updateTranslateURLBar(
+        tab: tab,
+        state: translateEnabled ? .available : .unavailable
+      )
+
+      return
+    }
+
+    // Let the user enable or disable translation via onboarding
+    let translateEnabled = await withCheckedContinuation { continuation in
+      delegate?.showTranslateOnboarding(tab: tab) { translateEnabled in
+        continuation.resume(returning: translateEnabled)
+      }
+    }
```

Comment:
if `delegate` is nil this continuation will never resume, you need to check if the delegate is nil before entering `withCheckedContinuation`